include/Dialect/Openfhe/IR/OpenfheOps.td

```diff
@@ -0,0 +1,54 @@
+#ifndef INCLUDE_DIALECT_OPENFHE_IR_OPENFHEOPS_TD_
+#define INCLUDE_DIALECT_OPENFHE_IR_OPENFHEOPS_TD_
+
+include "OpenfheDialect.td"
+include "OpenfheTypes.td"
+
+include "mlir/IR/BuiltinAttributes.td"
+include "mlir/IR/CommonTypeConstraints.td"
+include "mlir/IR/OpBase.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+
+class Openfhe_Op<string mnemonic, list<Trait> traits = []> :
+        Op<Openfhe_Dialect, mnemonic, traits> {
+  let assemblyFormat = [{
+    operands attr-dict `:` `(` type(operands) `)` `->` type(results)
+  }];
+  let cppNamespace = "::mlir::heir::openfhe";
+}
+
+def EncryptOp : Openfhe_Op<"encrypt", [Pure]> {
+  let arguments = (ins Openfhe_ServerKey:$serverKey, Openfhe_PlaintextType:$plaintext);
+  let results = (outs Openfhe_CiphertextType:$output);
+}
+
+def NegateOp : Openfhe_Op<"negate", [Pure, AllTypesMatch<["ciphertext", "output"]>]> {
+  let arguments = (ins Openfhe_ServerKey:$serverKey, Openfhe_CiphertextType:$ciphertext);
+  let results = (outs Openfhe_CiphertextType:$output);
+}
+
+def AddOp : Openfhe_Op<"add",[
+    Pure,
+    AllTypesMatch<["lhs", "rhs", "output"]>
+]> {
+  let arguments = (ins
+    Openfhe_ServerKey:$serverKey,
+    Openfhe_CiphertextType:$lhs,
+    Openfhe_CiphertextType:$rhs
+  );
+  let results = (outs Openfhe_CiphertextType:$output);
+}
+
+def SubOp : Openfhe_Op<"sub",[
+    Pure,
+    AllTypesMatch<["lhs", "rhs", "output"]>
+]> {
+  let arguments = (ins
+    Openfhe_ServerKey:$serverKey,
+    Openfhe_CiphertextType:$lhs,
+    Openfhe_CiphertextType:$rhs
+  );
+  let results = (outs Openfhe_CiphertextType:$output);
+}
+
```

Comment:
Please add a `MulOp` corresponding to [`EvalMult`](https://github.com/openfheorg/openfhe-development/blob/main/src/pke/include/cryptocontext.h#L1697)

I imagine we would also need [`RotateOp` ](https://github.com/openfheorg/openfhe-development/blob/main/src/pke/include/cryptocontext.h#L2083) here, which would lower from [`bgv.rotate`](https://github.com/google/heir/blob/5cdfcea42de3a872922919eebeda44a26424538d/include/Dialect/BGV/IR/BGVOps.td#L74)

The `CryptoContext` API has a lot of other higher level ops (squaring, sin/cos, evaluating a polynomial, etc.). I think we should not add those yet, but consider adding them in the future if they are more than a convenient wrapper.

Reply:
You may also want to consider adding ops that incorporate cleartexts (looks like they have overloaded their API for this case, e.g., [for EvalMult](https://github.com/openfheorg/openfhe-development/blob/main/src/pke/include/cryptocontext.h#L1942))

But I would be fine to wait until the need for it shows up in code generation.

Reply:
Applied!